File: scripts/crawler.js
#!/usr/bin/env node
/**
 * 城商行和农商行信用卡及外围系统投标信息爬虫
 */

const sqlite3 = require('sqlite3').verbose();
const path = require('path');
const fs = require('fs');
const { getAllSearchQueries, getTagsForQuery } = require('./search-config');

const DB_PATH = path.join(__dirname, '../data/news.db');

// 初始化数据库
function initDB() {
    return new Promise((resolve, reject) => {
        const dataDir = path.dirname(DB_PATH);
        if (!fs.existsSync(dataDir)) {
            fs.mkdirSync(dataDir, { recursive: true });
        }

        const db = new sqlite3.Database(DB_PATH, (err) => {
            if (err) reject(err);
            else resolve(db);
        });
    });
}

// 检查文章是否已存在
async function articleExists(db, url) {
    return new Promise((resolve, reject) => {
        db.get('SELECT id FROM articles WHERE url = ?', [url], (err, row) => {
            if (err) reject(err);
            else resolve(!!row);
        });
    });
}

// 保存文章
async function saveArticle(db, article) {
    const exists = await articleExists(db, article.url);
    if (exists) {
        console.log(`  [跳过] ${article.title.substring(0, 40)}...`);
        return false;
    }

    const sql = `
        INSERT INTO articles (title, summary, source, url, category, tags, publish_date)
        VALUES (?, ?, ?, ?, ?, ?, ?)
    `;
    
    return new Promise((resolve, reject) => {
        db.run(sql, [
            article.title,
            article.summary,
            article.source,
            article.url,
            article.category,
            JSON.stringify(article.tags),
            article.publish_date
        ], function(err) {
            if (err) {
                console.error(`  [错误] ${err.message}`);
                resolve(false);
            } else {
                console.log(`  [新增] ${article.title.substring(0, 40)}...`);
                resolve(true);
            }
        });
    });
}

// 处理搜索结果
async function processSearchResults(db, query, results) {
    console.log(`\n🔍 ${query}`);
    
    let added = 0;
    let skipped = 0;
    const tags = getTagsForQuery(query);
    
    for (const item of results) {
        const article = {
            title: item.title || '无标题',
            summary: item.snippet || item.summary || '',
            source: item.source || '未知来源',
            url: item.url,
            category: 'bid',
            tags: tags,
            publish_date: item.date || new Date().toISOString().split('T')[0]
        };
        
        const saved = await saveArticle(db, article);
        if (saved) {
            added++;
        } else {
            skipped++;
        }
    }
    
    console.log(`   新增: ${added}, 跳过: ${skipped}`);
    return { added, skipped };
}

// 运行爬虫
async function runCrawler(searchResults) {
    console.log('='.repeat(60));
    console.log('🏦 城商行和农商行信用卡及外围系统投标信息爬虫');
    console.log('='.repeat(60));
    console.log(`开始时间: ${new Date().toLocaleString()}`);
    console.log('');
    
    let db;
    let totalAdded = 0;
    let totalSkipped = 0;
    
    try {
        db = await initDB();
        
        // 确保表存在
        const schema = `
            CREATE TABLE IF NOT EXISTS articles (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                title TEXT NOT NULL,
                summary TEXT,
                content TEXT,
                source TEXT,
                url TEXT UNIQUE,
                category TEXT,
                tags TEXT,
                publish_date DATE,
                crawl_date DATETIME DEFAULT CURRENT_TIMESTAMP,
                is_read INTEGER DEFAULT 0,
                is_important INTEGER DEFAULT 0
            );
            CREATE INDEX IF NOT EXISTS idx_category ON articles(category);
            CREATE INDEX IF NOT EXISTS idx_publish_date ON articles(publish_date);
        `;
        await new Promise((resolve, reject) => {
            db.exec(schema, (err) => {
                if (err) reject(err);
                else resolve();
            });
        });
        
        // 处理搜索结果
        if (searchResults && Object.keys(searchResults).length > 0) {
            for (const [query, results] of Object.entries(searchResults)) {
                if (Array.isArray(results) && results.length > 0) {
                    const { added, skipped } = await processSearchResults(db, query, results);
                    totalAdded += added;
                    totalSkipped += skipped;
                }
            }
        } else {
            console.log('⚠️ 未提供搜索结果');
            console.log('请使用以下关键词通过 kimi_search 搜索:');
            const queries = getAllSearchQueries();
            queries.forEach((q, i) => {
                console.log(`  ${i + 1}. ${q}`);
            });
        }
        
        console.log('\n' + '='.repeat(60));
        console.log(`📊 总计: 新增 ${totalAdded} 条, 跳过 ${totalSkipped} 条`);
        console.log('='.repeat(60));
        
    } catch (err) {
        console.error('❌ 爬虫运行失败:', err);
        throw err;
    } finally {
        if (db) db.close();
    }
    
    return { totalAdded, totalSkipped };
}

// 主函数
async function main() {
    console.log('城商行和农商行信用卡及外围系统投标信息爬虫');
    console.log('');
    console.log('搜索关键词列表:');
    const queries = getAllSearchQueries();
    queries.forEach((q, i) => {
        console.log(`  ${i + 1}. ${q}`);
    });
    console.log('');
    console.log('使用方法:');
    console.log('1. 使用 OpenClaw kimi_search 工具搜索以上关键词');
    console.log('2. 将结果整理为 {query: [results]} 格式');
    console.log('3. 调用 runCrawler(searchResults) 保存到数据库');
}

// 如果直接运行此脚本
if (require.main === module) {
    main();
}

module.exports = { runCrawler, getAllSearchQueries };

File: scripts/crawler.py
#!/usr/bin/env python3
# 信用卡行业资讯平台 - 数据抓取模块
# 负责调用 Kimi Search 并解析结果

import json
import os
import sqlite3
from datetime import datetime

# 数据库路径
DB_PATH = os.path.join(os.path.dirname(__file__), '..', 'data', 'news.db')

# 搜索关键词配置
SEARCH_QUERIES = {
    'policy': [
        '国家金融监督管理总局 信用卡 通知 2026',
        '央行 信用卡 新规 政策 2026',
        '银保监会 信用卡 合规 2026',
    ],
    'news': [
        '信用卡 行业动态 2026',
        '银行信用卡 发卡量 报告 2026',
        '信用卡 市场 趋势 2026',
    ],
    'risk': [
        '城商行 农商行 信用卡 处罚 site:gov.cn 2026',
        '银行 信用卡 违规 罚款 site:nfra.gov.cn 2026',
        '银行 罚单 农商行 site:pbc.gov.cn 2026',
    ],
    'business': [
        '城商行 信用卡 系统 采购 2026',
        '农商行 信用卡 风控 招标 2026',
    ],
    'bid': [
        '城商行 信用卡 招标公告 2026',
        '农商行 信用卡 中标 2026',
        '城市商业银行 信用卡 采购 招标 2026',
    ],
}

def get_db():
    """获取数据库连接"""
    os.makedirs(os.path.dirname(DB_PATH), exist_ok=True)
    conn = sqlite3.connect(DB_PATH)
    conn.row_factory = sqlite3.Row
    return conn

def init_db():
    """初始化数据库表"""
    conn = get_db()
    conn.execute('''
        CREATE TABLE IF NOT EXISTS articles (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            title TEXT NOT NULL,
            summary TEXT,
            full_content TEXT,
            source TEXT,
            source_url TEXT,
            publish_date TEXT,
            crawl_time TEXT DEFAULT CURRENT_TIMESTAMP,
            primary_category TEXT,
            related_banks TEXT,
            is_duplicate INTEGER DEFAULT 0,
            duplicate_sources TEXT,
            penalized_institution TEXT,
            bank_type TEXT,
            penalty_authority TEXT,
            penalty_amount REAL,
            penalty_reason TEXT,
            penalty_document_no TEXT,
            project_name TEXT,
            procuring_entity TEXT,
            budget TEXT,
            bid_deadline TEXT,
            bid_status TEXT,
            winner TEXT,
            winner_amount TEXT,
            policy_no TEXT,
            issuing_authority TEXT,
            policy_level TEXT,
            is_key_policy INTEGER DEFAULT 0
        )
    ''')
    conn.commit()
    conn.close()

def insert_article(article):
    """插入文章"""
    conn = get_db()
    
    # 检查是否已存在
    exists = conn.execute(
        'SELECT COUNT(*) FROM articles WHERE title = ?',
        (article.get('title'),)
    ).fetchone()[0]
    
    if exists > 0:
        conn.close()
        return False
    
    conn.execute('''
        INSERT INTO articles (
            title, summary, source, source_url, publish_date, 
            primary_category, crawl_time
        ) VALUES (?, ?, ?, ?, ?, ?, ?)
    ''', (
        article.get('title'),
        article.get('summary'),
        article.get('source'),
        article.get('source_url'),
        article.get('publish_date'),
        article.get('primary_category'),
        datetime.now().strftime('%Y-%m-%d %H:%M:%S')
    ))
    conn.commit()
    conn.close()
    return True

def count_by_category():
    """统计分类数量"""
    conn = get_db()
    
    categories = ['policy', 'news', 'risk', 'business', 'bid']
    result = {'total': 0, 'byCategory': []}
    
    for cat in categories:
        count = conn.execute(
            f"SELECT COUNT(*) FROM articles WHERE primary_category=?",
            (cat,)
        ).fetchone()[0]
        result['byCategory'].append({'category': cat, 'count': count})
        result['total'] += count
    
    conn.close()
    return result

def format_for_feishu(stats):
    """格式化简报发送"""
    lines = ["🤖 每日市场资讯简报\n"]
    lines.append(f"📊 今日共 {stats['total']} 条资讯，分部如下：\n")
    
    cat_names = {
        'policy': '政策法规',
        'news': '行业资讯',
        'risk': '处罚数据',
        'business': '商机信息',
        'bid': '投标信息'
    }
    
    for item in stats['byCategory']:
        cat = item['category']
        count = item['count']
        name = cat_names.get(cat, cat)
        lines.append(f"- {name}：{count} 条\n")
    
    return ''.join(lines)

if __name__ == '__main__':
    # 测试用
    init_db()
    stats = count_by_category()
    print(json.dumps(stats, ensure_ascii=False))
    print('\n')
    print(format_for_feishu(stats))